### PR TITLE
Enable s3-provider support

### PR DESCRIPTION
**Pull request by ashimagarg27**
## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.
- [x] I have made sure existing UTs are not impacted
- [x] I have executed all necessary linter and made sure internal pipeline/travis is not broken
https://v3.travis.ibm.com/github/alchemy-containers/ibm-object-csi-driver/jobs/21976540

### CODE REVIEW

File: config/samples/csi_v1alpha1_ibmobjectcsi.yaml [hunk 1]
```diff
@@ -8,6 +8,8 @@ metadata:
     app.kubernetes.io/instance: ibm-object-csi
     release: v1.0.0
 spec:
+  s3Provider: "aws"
+  s3ProviderRegion: "us-east-2"
```
Comment (Bhagyashreek8) [inline config/samples/csi_v1alpha1_ibmobjectcsi.yaml:12]: We can remove the hard coded values

File: controllers/util/common/common.go [hunk 2]
```diff
@@ -331,17 +316,29 @@ func (ch *ControllerHelper) GetIBMCosSC() []string {
...
+	}
+}
+
+func (ch *ControllerHelper) SetS3ProviderEP() {
+	if ch.S3ProviderRegion == "" {
+		ch.CosEP = ""
+	}
+
+	if ch.S3Provider == constants.S3ProviderAWS {
+		ch.CosEP = fmt.Sprintf("https://s3.%s.amazonaws.com", ch.S3ProviderRegion)
```
Comment (Bhagyashreek8) [inline controllers/util/common/common.go:338]: move the URLs to constants

**Comment by ashimagarg27** [comment]
## Test Results
```
~ ❯ oc get all -n ibm-object-csi-operator                                                                                                                                                           ⎈ ms8-ai-syd/cqlgt6bs00avgplvrl20/admin 15:22:56
Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
NAME                                                              READY   STATUS    RESTARTS   AGE
pod/ibm-object-csi-operator-controller-manager-6489d6d649-whkkt   2/2     Running   0          118s

NAME                                                                 TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)    AGE
service/ibm-object-csi-operator-controller-manager-metrics-service   ClusterIP   172.21.77.57   <none>        8443/TCP   2m

NAME                                                         READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/ibm-object-csi-operator-controller-manager   1/1     1            1           119s

NAME                                                                    DESIRED   CURRENT   READY   AGE
replicaset.apps/ibm-object-csi-operator-controller-manager-6489d6d649   1         1         1       119s
```

```
~ ❯ oc describe ibmobjectcsi.objectdriver.csi.ibm.com/ibm-object-csi  -n ibm-object-csi-operator                                                                                                 6s ⎈ ms8-ai-syd/cqlgt6bs00avgplvrl20/admin 15:23:08
Name:         ibm-object-csi
Namespace:    ibm-object-csi-operator
Labels:       app.kubernetes.io/instance=ibm-object-csi
              app.kubernetes.io/name=ibm-object-csi
              release=v1.0.0
Annotations:  <none>
API Version:  objectdriver.csi.ibm.com/v1alpha1
Kind:         IBMObjectCSI
Metadata:
  Creation Timestamp:  2024-08-06T09:53:33Z
  Finalizers:
    ibmobjectcsi.objectdriver.csi.ibm.com
  Generation:        2
  Resource Version:  8328556
  UID:               72af7447-2023-4f04-aaab-811e011e58a4
Spec:
  Controller:
    Affinity:
      Node Affinity:
        Required During Scheduling Ignored During Execution:
          Node Selector Terms:
            Match Expressions:
              Key:       kubernetes.io/arch
              Operator:  In
              Values:
                amd64
      Pod Anti Affinity:
        Preferred During Scheduling Ignored During Execution:
          Pod Affinity Term:
            Label Selector:
              Match Expressions:
                Key:       app.kubernetes.io/component
                Operator:  In
                Values:
                  controller
            Topology Key:  topology.kubernetes.io/zone
          Weight:          100
          Pod Affinity Term:
            Label Selector:
              Match Expressions:
                Key:       app.kubernetes.io/component
                Operator:  In
                Values:
                  controller
            Topology Key:  topology.kubernetes.io/zone
          Weight:          100
    Image Pull Policy:     IfNotPresent
    Repository:            icr.io/ibm/ibm-object-csi-driver
    Resources:
      Limits:
        Cpu:     800m
        Memory:  400Mi
      Requests:
        Cpu:     40m
        Memory:  40Mi
    Tag:         v0.1.8
  Node:
    Image Pull Policy:  Always
    Repository:         icr.io/ibm/ibm-object-csi-driver
    Resources:
      Limits:
        Cpu:
        Memory:
      Requests:
        Cpu:
        Memory:
    Tag:             v0.1.8
  s3Provider:        aws
  s3ProviderRegion:  us-east-1
  Sidecars:
    Image Pull Policy:  IfNotPresent
    Name:               csi-node-driver-registrar
    Repository:         k8s.gcr.io/sig-storage/csi-node-driver-registrar
    Resources:
      Limits:
        Cpu:     40m
        Memory:  80Mi
      Requests:
        Cpu:            10m
        Memory:         20Mi
    Tag:                v2.6.3
    Image Pull Policy:  IfNotPresent
    Name:               csi-provisioner
    Repository:         k8s.gcr.io/sig-storage/csi-provisioner
    Resources:
      Limits:
        Cpu:     80m
        Memory:  160Mi
      Requests:
        Cpu:            20m
        Memory:         40Mi
    Tag:                v3.4.1
    Image Pull Policy:  IfNotPresent
    Name:               livenessprobe
    Repository:         k8s.gcr.io/sig-storage/livenessprobe
    Resources:
      Limits:
        Cpu:     20m
        Memory:  40Mi
      Requests:
        Cpu:     10m
        Memory:  10Mi
    Tag:         v2.9.0
Status:
  Controller Ready:  true
  Node Ready:        false
  Phase:             Creating
  Version:           1.0.0
Events:              <none>
```

**Comment by ashimagarg27** [comment]
```
```

**Comment by ashimagarg27** [comment]
```
~ ❯ oc get secret | grep cos-s3                                                                                                                                                                     ⎈ ms8-ai-syd/cqlgt6bs00avgplvrl20/admin 15:24:42
aws-s3fs-mount             cos-s3-csi-driver                     6      11s

~ ❯ oc describe secret aws-s3fs-mount                                                                                                                                                               ⎈ ms8-ai-syd/cqlgt6bs00avgplvrl20/admin 15:25:41
Name:         aws-s3fs-mount
Namespace:    default
Labels:       <none>
Annotations:  <none>

Type:  cos-s3-csi-driver

Data
====
bucketName:          6 bytes
cosEndpoint:         0 bytes
locationConstraint:  0 bytes
mounter:             4 bytes
secretKey:           40 bytes
accessKey:           20 bytes

~ ❯ oc get pvc                                                                                                                                                                                      ⎈ ms8-ai-syd/cqlgt6bs00avgplvrl20/admin 15:26:25
NAME             STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                       AGE
aws-s3fs-mount   Bound    pvc-c1fc2f87-9df9-4c8a-9579-d5583b8a8341   256Mi      RWO            ibm-object-storage-standard-s3fs   11s

~ ❯ oc get pv                                                                                                                                                                                       ⎈ ms8-ai-syd/cqlgt6bs00avgplvrl20/admin 15:26:49
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                    STORAGECLASS                       REASON   AGE
pvc-c1fc2f87-9df9-4c8a-9579-d5583b8a8341   256Mi      RWO            Delete           Bound    default/aws-s3fs-mount   ibm-object-storage-standard-s3fs            13s
```

```
~ ❯ oc get all | grep cos                                                                                                                                                                        5s ⎈ ms8-ai-syd/cqlgt6bs00avgplvrl20/admin 16:34:06
Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
pod/cos-csi-s3fs-644579b558-j4z6m   1/1     Running   0          19s
deployment.apps/cos-csi-s3fs   1/1     1            1           20s
replicaset.apps/cos-csi-s3fs-644579b558   1         1         1       20s

~ ❯ oc exec -it pod/cos-csi-s3fs-644579b558-j4z6m -- /bin/bash                                                                                                                                   4s ⎈ ms8-ai-syd/cqlgt6bs00avgplvrl20/admin 16:34:18
root@cos-csi-s3fs-644579b558-j4z6m:/# mount | grep s3
s3fs on /data1 type fuse.s3fs (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other)
root@cos-csi-s3fs-644579b558-j4z6m:/# ls /data1
root@cos-csi-s3fs-644579b558-j4z6m:/# cd /data1
root@cos-csi-s3fs-644579b558-j4z6m:/data1# touch test.txt
root@cos-csi-s3fs-644579b558-j4z6m:/data1# ls
test.txt
root@cos-csi-s3fs-644579b558-j4z6m:/data1# echo "testing" > test.txt
root@cos-csi-s3fs-644579b558-j4z6m:/data1# cat test.txt
testing
root@cos-csi-s3fs-644579b558-j4z6m:/data1#
```

**Review by Bhagyashreek8 — APPROVED** [review]
lgtm

**Review by nkkashyap — APPROVED** [review]
lgtm

**Pull request merged**